### ISSUE TITLE
[RFE] Expose "Streamer" extension, which is more practical at runtime than `HelperInterface::UploadData`

### ISSUE THREAD

**Issue by dzhdanNV** [NVIDIAGameWorks/NRI#34]
Draft with comments:
```c++
// © 2021 NVIDIA Corporation

#pragma once

NRI_NAMESPACE_BEGIN

NRI_FORWARD_STRUCT(Streamer);

// Could use an externally created "buffer", but in this case "growing under the hood" becomes impossible
NRI_STRUCT(StreamerDesc)
{
    //uint64_t capacity; // seems to be not needed
    NRI_NAME(MemoryLocation) memoryLocation; // UPLOAD or DEVICE_UPLOAD
    NRI_NAME(BufferUsageBits) usageBits;
    uint8_t framesInFlightNum; // needed to hide reallocation under the hood and avoid exposing capacity
};

NRI_STRUCT(BufferRangeUpdateRequestDesc)
{
    // Data to upload
    const void* data;
    uint64_t dataSize;

    // Destination (optional)
    NRI_NAME(Buffer)* dstBuffer;
    uint64_t dstBufferOffset;

    // Access?
    NRI_NAME(AccessBits) prevState; // potentially could assume UNKNOWN (unrecommended)
    NRI_NAME(AccessBits) nextState; // not all states can be transfered to in COPY and COMPUTE queues
};

NRI_STRUCT(TextureRegionUpdateRequestDesc)
{
    // Data to upload
    const void* data;
    uint64_t dataSize;
    uint32_t srcRowPitch;
    uint32_t srcSlicePitch;

    // Destination (mandatory)
    NRI_NAME(Texture)* dstTexture;
    const NRI_NAME(TextureRegionDesc)* dstRegionDesc;

    // Access?
    NRI_NAME(AccessAndLayout) prevState;
    NRI_NAME(AccessAndLayout) nextState;
};

NRI_STRUCT(StreamerInterface)
{
    NRI_NAME(Result) (NRI_CALL *CreateStreamer)(NRI_NAME_REF(Device) device, const NRI_NAME_REF(StreamerDesc) streamerDesc, NRI_NAME_REF(Streamer*) streamer);    
    void (NRI_CALL *DestroyStreamer)(NRI_NAME_REF(Streamer) streamer);

    // Add an update request to the queue (no work here)
    // These function return ring buffer offset
    uint64_t (NRI_CALL *EnqueueBufferRangeUpdateRequest)(NRI_NAME_REF(Streamer) streamer, const NRI_NAME_REF(BufferRangeUpdateRequestDesc) bufferRangeUpdateRequestDesc);
    uint64_t (NRI_CALL *EnqueueTextureRegionUpdateRequest)(NRI_NAME_REF(Streamer) streamer, const NRI_NAME_REF(TextureRegionUpdateRequestDesc) textureRegionUpdateRequestDesc);

    // Submit all gathered requests and reset the queue
    // The internal buffer can grow in this function
    // Doesn't requre WFI on a specific queue if a new buffer gets allocated immediately and destroying of the old buffer is postponed
    void (NRI_CALL *SubmitStreamerRequests)(NRI_NAME_REF(CommandBuffer) commandBuffer, NRI_NAME_REF(Streamer) streamer);

    // Needed if the buffer is explicitly used for rendering as IB, VB, CB
    // IMPORTANT: valid only after "Submit"
    // IMPORTANT: creating and caching "views" on this buffer is unrecommended
    NRI_NAME(Buffer*) (NRI_CALL *GetStreamerBuffer)(NRI_NAME_REF(Streamer) streamer);

    /*
    Not needed with "enqueue / submit" logic:
    uint64_t (NRI_CALL *GetStreamerCapacity)(NRI_NAME_REF(Streamer) streamer);
    void (NRI_CALL *ChangeStreamerCapacity)(NRI_NAME_REF(Streamer) streamer, uint64_t capacity);
    */
};

NRI_NAMESPACE_END
```
Such design allows to hide capacity, but "puts" all CPU-side copy operations in one place, i.e. 1 thread. It's not good, but at the same time other threads can have own `Streamer` objects.

**Comment by dzhdanNV** [comment]
@vertver FYI. Feel free to comment.

**Comment by dzhdanNV** [comment]
Thanks for enthusiasm.  I'm planning to implement it after improving barriers #35.

**Comment by dzhdanNV** [comment]
I implemented modern barriers functionality. Now I'm planning to work on this...

**Comment by vertver** [comment]
Nice. How about I handle this extension and you implement the piping library? It seems like extra functionality, and at the same time I've already implemented all the necessary code for it.

**Comment by dzhdanNV** [comment]
Please, better start to work on Pipeline Library (at least interface). I will handle streaming. But I will be on vacation next week. I haven't read PL docs much, but it will be tricky - IIRC, D3D12 offers lots of functionality in this area, while VK only 1 new function :)

**Comment by vertver** [comment]
Ok, got you.

**Comment by dzhdanNV** [comment]
Currently working on a better API design with a colleague. The idea to move all barriers out of the stream to unblock:
- better grouping
- global memory barriers
- "no" barriers approach if used in "another" ECL (queue submit call)

**Comment by vertver** [comment]
Sounds cool, I've been waiting for this change!

**Comment by dzhdanNV** [comment]
Final version of the interface (until some problems found):
```c++
// © 2021 NVIDIA Corporation

#pragma once

NRI_NAMESPACE_BEGIN

NRI_FORWARD_STRUCT(Streamer);

NRI_STRUCT(StreamerDesc)
{    
    NRI_NAME(MemoryLocation) memoryLocation; // UPLOAD or DEVICE_UPLOAD
    NRI_NAME(BufferUsageBits) usageBits; // CONSTANT_BUFFER, INDEX_BUFFER or VERTEX_BUFFER
    uint8_t framesInFlightNum;
};

NRI_STRUCT(BufferRangeUpdateRequestDesc)
{
    // Data to upload
    const void* data;
    uint64_t dataSize;

    // Destination (optional)
    NRI_NAME(Buffer)* dstBuffer;
    uint64_t dstBufferOffset;
};

NRI_STRUCT(TextureRegionUpdateRequestDesc)
{
    // Data to upload
    const void* data;
    uint64_t dataSize;
    uint32_t srcRowPitch;
    uint32_t srcSlicePitch;

    // Destination (mandatory)
    NRI_NAME(Texture)* dstTexture;
    NRI_NAME(TextureRegionDesc) dstRegionDesc;
};

NRI_STRUCT(StreamerInterface)
{
    NRI_NAME(Result) (NRI_CALL *CreateStreamer)(NRI_NAME_REF(Device) device, const NRI_NAME_REF(StreamerDesc) streamerDesc, NRI_NAME_REF(Streamer*) streamer);
    void (NRI_CALL *DestroyStreamer)(NRI_NAME_REF(Streamer) streamer);

    // Add an update request to the queue. Functions return the ring buffer offset and don't invoke any GPU work
    uint64_t (NRI_CALL *EnqueueBufferRangeUpdateRequest)(NRI_NAME_REF(Streamer) streamer, const NRI_NAME_REF(BufferRangeUpdateRequestDesc) bufferRangeUpdateRequestDesc);
    uint64_t (NRI_CALL *EnqueueTextureRegionUpdateRequest)(NRI_NAME_REF(Streamer) streamer, const NRI_NAME_REF(TextureRegionUpdateRequestDesc) textureRegionUpdateRequestDesc);

    // Add gathered requests to a command buffer and reset the streamer queue. Barriers are externally controlled
    void (NRI_CALL *CmdUpload)(NRI_NAME_REF(CommandBuffer) commandBuffer, NRI_NAME_REF(Streamer) streamer);

    // Valid only after "CmdUpload", where the internal buffer can be changed
    NRI_NAME(Buffer*) (NRI_CALL *GetStreamerBuffer)(NRI_NAME_REF(Streamer) streamer);
};

NRI_NAMESPACE_END
```

**Comment by dzhdanNV** [comment]
Implemented. Interface has been changed during development.
